backends/nats/nats.go

```diff
@@ -0,0 +1,486 @@
+	// Not loaded from YAML
+	Logger logr.Logger `yaml:"-"`
+}
+
+type Backend struct {
+	opt        Options
+	nc         *nats.Conn
+	log        logr.Logger
+	markerName string
+}
+
+// Check TLS params are available
+func (o Options) checkTLS() error {
+	if o.NatsTLSRootCA != "" {
+		err := fileExistsAndIsReadable(o.NatsTLSRootCA)
+		if err != nil {
+			return err
+		}
+	}
+	if (o.NatsTLSClientCert != "" && o.NatsTLSClientKey == "") || (o.NatsTLSClientKey != "" && o.NatsTLSClientCert == "") {
+		return errors.New("you appear to want to use client TLS certs but have not provided either cert or key")
+	}
+	if o.NatsTLSClientCert != "" {
+		err := fileExistsAndIsReadable(o.NatsTLSClientCert)
+		if err != nil {
+			return err
+		}
+	}
+	if o.NatsTLSClientKey != "" {
+		err := fileExistsAndIsReadable(o.NatsTLSClientKey)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Check some credentials have been supplied and set the flag in the options struct
+func (o Options) checkCredentialsAvailability() error {
+	// Found a token
+	if o.NatsToken != "" {
+		o.internalUseAuthType = token
+		return nil
+	}
+	// Found a username
+	if o.NatsUsername != "" && o.NatsPassword != "" {
+		o.internalUseAuthType = username
+		return nil
+	}
+	if (o.NatsUsername != "" && o.NatsPassword == "") || (o.NatsPassword != "" && o.NatsUsername == "") {
+		o.internalUseAuthType = invalid
+		return errors.New("you appear to want to authenticate using username and password but have not supplied a username or a password ")
+	}
+	// Nkey specified
+	if o.NatsNkeySeedFilePath != "" {
+		err := fileExistsAndIsReadable(o.NatsNkeySeedFilePath)
+		if err != nil {
+			o.internalUseAuthType = invalid
+			return err
+		}
+		if o.NatsJWTKeyFilePath == "" {
+			o.internalUseAuthType = nkey
+			return nil
+		}
+		err = fileExistsAndIsReadable(o.NatsJWTKeyFilePath)
+		if err != nil {
+			o.internalUseAuthType = invalid
+			return err
+		}
+		o.internalUseAuthType = nkeyJwt
+		return nil
+	}
+	// JWT only
+	if o.NatsJWTKeyFilePath != "" {
+		err := fileExistsAndIsReadable(o.NatsJWTKeyFilePath)
+		if err != nil {
+			o.internalUseAuthType = invalid
+			return err
+		}
+		o.internalUseAuthType = jwt
+		return nil
+	}
+	// Credentials specified
+	if o.NatsCredentialsFilePath != "" {
+		err := fileExistsAndIsReadable(o.NatsCredentialsFilePath)
+		if err != nil {
+			o.internalUseAuthType = invalid
+			return err
+		}
+		o.internalUseAuthType = credentials
+		return nil
+	}
+	o.internalUseAuthType = invalid
+	return errors.New("unable to detect suitable credentials")
+}
+
+// prependGlobalPrefix prepends the GlobalPrefix to the name/prefix
+// passed as input
+func (b *Backend) prependGlobalPrefix(name string) string {
+	return b.opt.GlobalPrefix + name
+}
+
+// setGlobalPrefix updates the global prefix in b and the cached marker name,
+// so it can be dynamically changed in tests.
+func (b *Backend) setGlobalPrefix(prefix string) {
+	b.opt.GlobalPrefix = prefix
+}
```

Comment:
Prefixes made sense in the context of S3 buckets, but perhaps not in the context of NATS buckets that can easily be created on the fly.

Reply:
Ultimatley the fundamental problem if if I was testing the backend against `tester/DoBackendTests` then I had little choice to implement prefix support, otherwise `DoBackendTests` would complain.